test/pgo/IntegrationTestingUtils.java

```diff
@@ -26,6 +47,97 @@ public static void expungeFile(File file) {
 			System.err.println("WARNING: could not delete file "+file+"; check your temp folder");
 		}
 	}
+
+	public static void testCompileExpression(PGoTLAExpression result, List<KeyValue> vars,
+											 List<String> expected, Boolean panic) throws IOException {
+		Path tempDirPath = Files.createTempDirectory("pgotest");
+		File tempDir = tempDirPath.toFile();
+		Path generatedCodePath = tempDirPath.resolve("Test.tla");
+		Path generatedConfigPath = tempDirPath.resolve("config.json");
+
+		Path compiledOutputPath = tempDirPath.resolve("test.go");
+		try{
+			// generate test TLA+ file
+			try(BufferedWriter w = Files.newBufferedWriter(generatedCodePath);
+				IndentingWriter out = new IndentingWriter(w);){
+				out.write("---- MODULE Test ----");
+				out.newLine();
+				out.write("EXTENDS Sequences, Integers");
+				out.newLine();
+
+				out.write("(* --algorithm Test {");
+				out.newLine();
+				try(IndentingWriter.Indent i_ = out.indent()){
+					if(!vars.isEmpty()) {
+						out.write("variables ");
+						try(IndentingWriter.Indent i2_ = out.indentToPosition()){
+							for(KeyValue var : vars) {
+								out.write(var.getKey());
+								out.write(" = ");
+								var.getValue().accept(new PGoTLAExpressionFormattingVisitor(out));
+								out.write(";");
+								out.newLine();
+							}
+						}
+					}
+					out.write("{");
+					out.newLine();
+					try(IndentingWriter.Indent i2_ = out.indent()){
+						out.write("print ");
+						result.accept(new PGoTLAExpressionFormattingVisitor(out));
+					}
+					out.newLine();
+					out.write("}");
+				}
+				out.newLine();
+				out.write("}");
+				out.newLine();
+				out.write("*)");
+
+				out.newLine();
+				out.write("====");
+			}
+
+			// generate config file
+			try(BufferedWriter w = Files.newBufferedWriter(generatedConfigPath)){
+				JSONObject config = new JSONObject();
+
+				JSONObject build = new JSONObject();
+				build.put("output_dir", tempDirPath.toString());
+				build.put("dest_file", "test.go");
+				config.put("build", build);
+
+				JSONObject networking = new JSONObject();
+				networking.put("enabled", false);
+				config.put("networking", networking);
+
+				config.write(w);
+			}
+
+			// invoke the compiler
+			PGoMain.main(new String[] {
+					"-c",
+					generatedConfigPath.toString(),
+					generatedCodePath.toString(),
+			});
+
+			// display the compiled code for inspection
+			Files.lines(compiledOutputPath).forEach(line -> {
+				System.out.println("source: "+line);
+			});
+
+			System.out.println("Got here ok");
+
+			// try to run the compiled Go code, check that it prints the right thing
+			if (!panic) {
```

Comment:
Thought: for a more meaningful parameterization of this function, we could replace the `panic` argument with a `Consumer<Path>` that you might call like `runner.accept(compiledOutputPath);`.

Example callsite for this function with that change:
```
IntegrationTestingUtils.testCompileExpression(result, vars, compiledOutputPath ->
        IntegrationTestingUtils.testRunGoCode(compiledOutputPath, expected);
```

Reply:
Good idea, that will make it very easy to add new testing functionality.

Reply:
Cleaned up by using this interface definition in IntegrationTestingUtils:
```
public interface TestRunner<T> {
	public void run(T t) throws IOException;
}
```